cmd/car/compile.go

```diff
@@ -0,0 +1,446 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"regexp"
+	"strings"
+	"unicode/utf8"
+
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	carv1 "github.com/ipld/go-car"
+	"github.com/ipld/go-car/util"
+	carv2 "github.com/ipld/go-car/v2"
+	"github.com/ipld/go-car/v2/blockstore"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/codec/dagjson"
+	"github.com/ipld/go-ipld-prime/datamodel"
+	"github.com/ipld/go-ipld-prime/linking"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/node/basicnode"
+	"github.com/ipld/go-ipld-prime/storage/memstore"
+	"github.com/urfave/cli/v2"
+)
+
+// Compile is a command to translate between a human-debuggable patch-like format and a car file.
+func CompileCar(c *cli.Context) error {
+	var err error
+	inStream := os.Stdin
+	if c.Args().Len() >= 1 {
+		inStream, err = os.Open(c.Args().First())
+		if err != nil {
+			return err
+		}
+	}
+
+	//parse headers.
+	br := bufio.NewReader(inStream)
+	header, _, err := br.ReadLine()
+	if err != nil {
+		return err
+	}
+
+	v2 := strings.Contains(string(header), "--v2")
+	trimH := strings.TrimSpace(string(header))
+	headerParts := strings.Split(trimH, " ")
+	carName := headerParts[len(headerParts)-1]
+
+	roots := make([]cid.Cid, 0)
+	for {
+		peek, err := br.Peek(4)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
+		if bytes.Equal(peek, []byte("--- ")) {
+			break
+		}
+		rootLine, _, err := br.ReadLine()
+		if err != nil {
+			return err
+		}
+		if strings.HasPrefix(string(rootLine), "root ") {
```

Comment:
ok, so this is is not exhaustive, if it doesn't match `root ` then it continues to loop and drop `rootLine`. So is this where we get potential backward compatibility of additional v2 features that we can insert with `---`, and also the `--foobar` arguments in the header?

Reply:
(continuing my thought below about v2 features, where I started my comments on debug first before moving up here to compile)

Reply:
yes, other meta info would go as lines in this section.